symengine/infinity.cpp

```diff
@@ -238,4 +238,295 @@ inline RCP<const Infty> infty(const RCP<const Number> &direction)
+            return one;
+        } else if (s.is_negative()) {
+            return minus_one;
+        } else {
+            throw DomainError(
+                "coth(x) is not defined for x = Complex Infinity");
+        }
+    }
+    virtual RCP<const Basic> asinh(const Basic &x) const override
+    {
+        SYMENGINE_ASSERT(is_a<Infty>(x))
+        const Infty &s = down_cast<const Infty &>(x);
+        if (s.is_positive()) {
+            return Inf;
+        } else if (s.is_negative()) {
+            return NegInf;
+        } else {
+            throw DomainError(
+                "asinh(x) is not defined for x = Complex Infinity");
+        }
+    }
+    virtual RCP<const Basic> acosh(const Basic &x) const override
+    {
+        SYMENGINE_ASSERT(is_a<Infty>(x))
+        const Infty &s = down_cast<const Infty &>(x);
+        if (s.is_positive() or s.is_negative()) {
+            return Inf;
+        } else {
+            throw DomainError(
+                "acosh(x) is not defined for x = Complex Infinity");
+        }
+    }
+    virtual RCP<const Basic> acsch(const Basic &x) const override
+    {
+        SYMENGINE_ASSERT(is_a<Infty>(x))
+        const Infty &s = down_cast<const Infty &>(x);
+        if (s.is_positive() or s.is_negative()) {
+            return zero;
+        } else {
+            throw DomainError(
+                "acsch(x) is not defined for x = Complex Infinity");
+        }
+    }
+    virtual RCP<const Basic> atanh(const Basic &x) const override
+    {
+        SYMENGINE_ASSERT(is_a<Infty>(x))
+        const Infty &s = down_cast<const Infty &>(x);
+        if (s.is_positive()) {
+            return mul(minus_one, div(mul(pi, I), integer(2)));
+        } else if (s.is_negative()) {
+            return div(mul(pi, I), integer(2));
+        } else {
+            throw DomainError(
+                "atanh(x) is not defined for x = Complex Infinity");
+        }
+    }
+    virtual RCP<const Basic> acoth(const Basic &x) const override
+    {
+        SYMENGINE_ASSERT(is_a<Infty>(x))
+        const Infty &s = down_cast<const Infty &>(x);
+        if (s.is_positive() or s.is_negative()) {
+            return zero;
+        } else {
+            throw DomainError(
+                "acoth(x) is not defined for x = Complex Infinity");
+        }
+    }
+    virtual RCP<const Basic> abs(const Basic &x) const override
+    {
+        SYMENGINE_ASSERT(is_a<Infty>(x))
+        return Inf;
+    }
+    virtual RCP<const Basic> log(const Basic &x) const override
+    {
+        SYMENGINE_ASSERT(is_a<Infty>(x))
+        const Infty &s = down_cast<const Infty &>(x);
+        if (s.is_positive()) {
+            return Inf;
+        } else if (s.is_negative()) {
+            return Inf;
+        } else {
+            return ComplexInf;
+        }
+    }
+    virtual RCP<const Basic> gamma(const Basic &x) const override
+    {
+        SYMENGINE_ASSERT(is_a<Infty>(x))
+        throw NotImplementedError("Not Implemented.");
```

Comment:
@isuruf for  gamma(NegInf) and gamma(ComplexInf) should i return ComplexInf 
Ref : https://github.com/symengine/symengine/blob/master/symengine/functions.cpp#L2419-L2427